File: ios/DealFlow/Models/Deal.swift
import Foundation
import SwiftUI

// MARK: - Deal Stage
enum DealStage: String, Codable, CaseIterable, Identifiable {
    case sourced = "Sourced"
    case investorTargeting = "Investor Targeting"
    case diligence = "Diligence"
    case termSheet = "Term Sheet"
    case negotiation = "Negotiation"
    case closed = "Closed"
    case passed = "Passed"

    var id: String { rawValue }

    var color: Color {
        switch self {
        case .sourced: return .blue
        case .investorTargeting: return .purple
        case .diligence: return .orange
        case .termSheet: return Color(red: 0.8, green: 0.6, blue: 0.0)
        case .negotiation: return .teal
        case .closed: return .green
        case .passed: return .red
        }
    }

    var isActive: Bool {
        return self != .closed && self != .passed
    }

    var emoji: String {
        switch self {
        case .sourced: return "🔍"
        case .investorTargeting: return "🎯"
        case .diligence: return "📋"
        case .termSheet: return "📄"
        case .negotiation: return "🤝"
        case .closed: return "✅"
        case .passed: return "❌"
        }
    }
}

// MARK: - Deal Model
struct Deal: Identifiable, Codable, Equatable {
    var id: String
    var companyName: String
    var stage: DealStage
    var raiseAmount: Double?
    var valuation: Double?
    var sector: String?
    var dealOwner: String?
    var website: String?
    var notes: String?
    var userId: String?
    var workspaceId: String?
    var feePct: Double?
    var timelineToClose: String?
    var memo: String?
    var contacts: [DealContact]?
    var coInvestors: [String]?
    var buyers: [Buyer]?
    var metrics: DealMetrics?
    var score: DealScore?
    var createdAt: String?
    var updatedAt: String?

    enum CodingKeys: String, CodingKey {
        case id
        case companyName = "company_name"
        case stage
        case raiseAmount = "raise_amount"
        case valuation
        case sector
        case dealOwner = "deal_owner"
        case website
        case notes
        case userId = "user_id"
        case workspaceId = "workspace_id"
        case feePct = "fee_pct"
        case timelineToClose = "timeline_to_close"
        case memo
        case contacts
        case coInvestors = "co_investors"
        case buyers
        case metrics
        case score
        case createdAt = "created_at"
        case updatedAt = "updated_at"
    }

    static func == (lhs: Deal, rhs: Deal) -> Bool {
        lhs.id == rhs.id
    }

    var estimatedFee: Double? {
        guard let raise = raiseAmount, let pct = feePct else { return nil }
        return raise * (pct / 100.0)
    }
}

// MARK: - Supporting Models
struct DealContact: Codable, Identifiable, Equatable {
    var id: String
    var name: String
    var title: String?
    var email: String?
    var notes: String?

    init(id: String = UUID().uuidString, name: String, title: String? = nil,
         email: String? = nil, notes: String? = nil) {
        self.id = id
        self.name = name
        self.title = title
        self.email = email
        self.notes = notes
    }
}

struct Buyer: Codable, Identifiable, Equatable {
    var id: String
    var name: String
    var type: String?
    var rationale: String?

    init(id: String = UUID().uuidString, name: String,
         type: String? = nil, rationale: String? = nil) {
        self.id = id
        self.name = name
        self.type = type
        self.rationale = rationale
    }
}

struct DealMetrics: Codable, Equatable {
    var revenue: Double?
    var ebitda: Double?
    var ebitdaMargin: Double?
    var arr: Double?
    var growthRate: Double?
    var grossMargin: Double?
    var employees: Int?
    var founded: String?
    var other: String?

    enum CodingKeys: String, CodingKey {
        case revenue, ebitda
        case ebitdaMargin = "ebitda_margin"
        case arr
        case growthRate = "growth_rate"
        case grossMargin = "gross_margin"
        case employees, founded, other
    }
}

struct DealScore: Codable, Equatable {
    var team: Int?
    var market: Int?
    var traction: Int?
    var terms: Int?
    var overall: Int?
    var rationale: String?
}

// MARK: - Sector Options
enum DealSector: String, CaseIterable {
    case technology = "Technology"
    case gaming = "Gaming"
    case mediaEntertainment = "Media & Entertainment"
    case hospitality = "Hospitality"
    case sports = "Sports"
    case healthcare = "Healthcare"
    case consumer = "Consumer"
    case finance = "Finance"
    case realEstate = "Real Estate"
    case other = "Other"
}

File: mobile/types/deal.ts
export interface Task {
  id: string;
  text: string;
  done: boolean;
  due?: string;
}

export interface Contact {
  id: string;
  name: string;
  title?: string;
  firm?: string;
  email?: string;
  phone?: string;
  notes?: string;
}

export interface Activity {
  id: string;
  type: "call" | "email" | "meeting" | "note";
  note: string;
  date: string;
}

export interface Buyer {
  id: string;
  name: string;
  status: "Active" | "Passed";
  email?: string;
  phone?: string;
  firm?: string;
  notes?: string;
  emailed?: boolean;
}

export interface Document {
  id: string;
  name: string;
  url: string;
  size?: number;
  type?: string;
  uploaded_at?: string;
}

export interface Metrics {
  revenue?: string;
  ebitda?: string;
  ebitda_margin?: string;
  arr?: string;
  growth_rate?: string;
  gross_margin?: string;
  employees?: string;
  founded?: string;
  other?: string;
}

export interface BuyerUniverse {
  name: string;
  type: "PE" | "Strategic" | "Family Office" | "Growth Equity";
  thesis?: string;
  website?: string;
  contact_name?: string;
  contact_title?: string;
  email?: string;
  phone?: string;
  notes?: string;
}

export interface Deal {
  id: string;
  created_at: string;
  updated_at: string;
  company_name: string;
  stage: string;
  sector?: string;
  raise_amount?: number;
  valuation?: number;
  fee_pct?: number;
  deal_owner?: string;
  website?: string;
  notes?: string;
  memo?: string;
  expected_close_date?: string;
  timeline_to_close?: string;
  user_id?: string;
  workspace_id?: string;
  // JSONB fields
  tasks?: Task[];
  activity_log?: Activity[];
  buyers?: Buyer[];
  contacts?: Contact[];
  documents?: Document[];
  metrics?: Metrics;
  score?: Record<string, unknown>;
  buyer_universe?: BuyerUniverse[];
  comments?: unknown[];
  shared_with?: string[];
}
